### DISCUSSION TITLE
Self one-to-one relation only allows a single row without a relation due to unique constraint?

### DISCUSSION THREAD

**Discussion by hoopyfroody** [prisma/prisma#12156] [Q&A] [answered]
If I make one-to-one self-relation like this:

```
model Constant {
  id            Int       @id @default(autoincrement())
  successor     Constant? @relation("ConstantHistory")
  predecessor   Constant? @relation("ConstantHistory", fields: [predecessorId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  predecessorId Int?
}
```
I end up with the `predecessorId` being set unique in the database schema, which does not allow me to have more than 1 Constant without a predecessor. I was hoping to use this construction to be able to have several linked lists:

(predecessorId values)
- null->1->2->5
- null->3->4

but this added constraint makes this impossible.

Do I need to change my approach? I guess I could make it a one-to-many relation and use it as a one-to-one, but that's a bit of a hack.

**Comment by austincrim** [comment]
Hey @mattixittam 👋 ,

I recreated your schema and I was able to successfully create multiple `Constant`s with a `null` `predecessorId`.
```ts
[
  { id: 1, predecessorId: null },
  { id: 2, predecessorId: 1 },
  { id: 3, predecessorId: null },
  { id: 4, predecessorId: 3 }
]
```

What database provider are you using? Which version of Prisma?

Thanks!

**Reply by hoopyfroody** [reply]
I am using `sqlserver` and prisma `3.10.0`

This is what happens when I try to insert a second record without a predecessorId in Prisma Studio:

![image](https://user-images.githubusercontent.com/2349560/156877206-4abfefaa-347f-4725-8d23-d16623aeffce.png)

This is the migration generated by Prisma:

```sql
-- CreateTable
CREATE TABLE [dbo].[Constant] (
    [id] INT NOT NULL IDENTITY(1,1),
    [predecessorId] INT,
    CONSTRAINT [Constant_pkey] PRIMARY KEY ([id]),
    CONSTRAINT [Constant_predecessorId_key] UNIQUE ([predecessorId])
);

-- AddForeignKey
ALTER TABLE [dbo].[Constant] ADD CONSTRAINT [Constant_predecessorId_fkey] FOREIGN KEY ([predecessorId]) REFERENCES [dbo].[Constant]([id]) ON DELETE NO ACTION ON UPDATE NO ACTION;

**Reply by austincrim** [reply]
It looks like SQL Server does not allow multiple `NULL` values for a unique constraint, while other DB's do (I was using Postgres, which explains why it worked for me). 

You can read more in [this StackOverflow thread](https://dba.stackexchange.com/questions/80514/why-does-a-unique-constraint-allow-only-one-null) and you might be able to find some workarounds there as well.

**Reply by hoopyfroody** [reply, marked as answer]
Yeah, I found that thread. I think the workaround there should be the way Prisma does it, to make the behavior equal between database connectors.

So I have opened up an issue to that effect: https://github.com/prisma/prisma/issues/12181

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
